### PR TITLE
add usb host device test scenarios

### PR DESCRIPTION
**Pull request by brucepatricklee**
1. cold-plug/unplug usb host device to/from vm
2. hot-plug/unplug usb host device to/from vm

Signed-off-by: Jin Li <jil@redhat.com>

### CODE REVIEW

**Comment by qiankehan** [comment]
Hi, @brucepatricklee 
I have tried some of the cases but they are failed:
JOB ID     : b6847e3375c255a0a37a19a76ac8631d24853e4a
JOB LOG    : /root/avocado/job-results/job-2019-04-30T17.25-b6847e3/job.log
 (1/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb30_controller.nec_xhci_model_hub.pci-root: FAIL (33.29 s)
 (2/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb30_controller.qemu_xhci_model.pci-root: FAIL (19.15 s)
 (3/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.nec_xhci_model_hub.pci-root: FAIL (15.65 s)
 (4/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.qemu_xhci_model.pci-root: FAIL (16.70 s)

The error in log:
2019-04-30 17:28:08,740 process          L0642 INFO | Command 'tar -zcf /root/avocado/job-results/job-2019-04-30T17.25-b6847e3/test-results/4-type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.qemu_xhci_model.pci-root/libvirtd.tar.gz -P /root/avocado/job-results/job-2019-04-30T17.25-b6847e3/test-results/4-type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.qemu_xhci_model.pci-root/libvirtd.log' finished with 0 after 0.0680530071259s
2019-04-30 17:28:08,741 stacktrace       L0041 ERROR|
2019-04-30 17:28:08,741 stacktrace       L0044 ERROR| Reproduced traceback from: /root/avocado/avocado/core/test.py:864
2019-04-30 17:28:08,742 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-04-30 17:28:08,742 stacktrace       L0047 ERROR|   File "/root/avocado-vt/avocado_vt/test.py", line 297, in runTest
2019-04-30 17:28:08,742 stacktrace       L0047 ERROR|     raise self.__status  # pylint: disable=E0702
2019-04-30 17:28:08,742 stacktrace       L0047 ERROR| TestFail: no available usb device in host
2019-04-30 17:28:08,742 stacktrace       L0048 ERROR|

I noticed that the cases have keyword **passthrough**. So are there some prerequisites for these cases? What else I need to prepare excluding the avocado-vt env? Do I need to prepare usb device in host  before the test run?
It is better to add this info in the code comments or commit msgs.

**Comment by brucepatricklee** [comment]
@qiankehan the log shows "no available usb device in host", you should plug a usb device before testing.
update the code comments to include this message

File: libvirt/tests/src/usb_device.py [hunk 1]
```diff
@@ -216,18 +228,37 @@ def usb_device_check(session, src_host=None):
...
 
+        # hotplug usb passthrough device
+        if passthrough and hotplug:
+            for hostdev in hostdev_list:
+                virsh.attach_device(vm_name, hostdev.xml, flagstr="--live",
+                                    debug=True, ignore_status=False)
+                utils_misc.wait_for(lambda: not session.cmd_status(
+                                     "lsusb | grep {}".format(hostdev.source.product_id)), 10)
+            vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+            logging.debug("vmxml after attaching {}".format(vmxml))
```
Comment (chloerh) [inline libvirt/tests/src/usb_device.py]: We could try this 
logging.debug("vmxml after attaching %s", vmxml)
it should also work.
Just suggesting

File: libvirt/tests/src/usb_device.py [hunk 2]
```diff
@@ -216,18 +228,37 @@ def usb_device_check(session, src_host=None):
...
+
         # check usb device
         usb_device_check(session, src_host)
 
         if passthrough:
             # detach usb passthrough device from vm
             hostdevs = vmxml.get_devices('hostdev')
-            logging.debug("detach usb device {}".format(hostdevs))
+            if coldunplug:
+                vm.shutdown()
```
Comment (chloerh) [inline libvirt/tests/src/usb_device.py]: There's also a 'wait_for_shutdown', it looks better, doesn't it?

File: libvirt/tests/src/usb_device.py [hunk 3]
```diff
@@ -216,18 +228,37 @@ def usb_device_check(session, src_host=None):
...
+                    if coldunplug:
+                        vmxml.del_device(dev)
+                    else:
+                        virsh.detach_device(vm_name, dev.xml, flagstr="--live", debug=True, ignore_status=False)
 
             # check the hostdev element in xml after detaching
+            if coldunplug:
+                vmxml.sync()
+                vm.start()
+                vm.wait_for_login(timeout=start_timeout)
```
Comment (chloerh) [inline libvirt/tests/src/usb_device.py]: If we just want to wait until vm is fully started, the session that 'wait_for_login' created should be closed right after that.

File: libvirt/tests/src/usb_device.py (hunk 2 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: no, the shutdown() will use wait_for_shutdown to check the vm shutting down process.
    def shutdown(self):
        """
        Shuts down this VM.
        """
        try:
            if self.state() != 'shut off':
                virsh.shutdown(self.name, uri=self.connect_uri)
            if self.wait_for_shutdown():
                logging.debug("VM %s shut down", self.name)
                self.cleanup_serial_console()
                return True
            else:
                logging.error("VM %s failed to shut down", self.name)
                return False
        except process.CmdError:
            logging.error("VM %s failed to shut down", self.name)
            return False

File: libvirt/tests/src/usb_device.py (hunk 3 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: good point, updated with wait_for_login().close()

File: libvirt/tests/src/usb_device.py (hunk 1 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: Compared with printf-style formatting, format function is a better choice which Python suggest people to use. So there is no need to change

File: libvirt/tests/src/usb_device.py (hunk 2 above)
Comment (chloerh) [reply libvirt/tests/src/usb_device.py]: :ok

Comment (chloerh) [reply libvirt/tests/src/usb_device.py]: :ok:

**Comment by brucepatricklee** [comment]
@qiankehan code updated, please help to review

**Comment by qiankehan** [comment]
@brucepatricklee 
I run your cases via a host with usb keyborad but the result is failed:
> 2019-05-06 13:44:24,650 stacktrace       L0041 ERROR|
2019-05-06 13:44:24,650 stacktrace       L0044 ERROR| Reproduced traceback from: /root/avocado/avocado/core/test.py:864
2019-05-06 13:44:24,678 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-05-06 13:44:24,678 stacktrace       L0047 ERROR|   File "/root/avocado-vt/avocado_vt/test.py", line 297, in runTest
2019-05-06 13:44:24,678 stacktrace       L0047 ERROR|     raise self.__status  # pylint: disable=E0702
2019-05-06 13:44:24,679 stacktrace       L0047 ERROR| TestFail: no available usb storage device

Running cases with usb disk will be a success:
```
JOB ID     : b9bcf5f952fb3ef60d677b5d7be1546075f085b1
JOB LOG    : /root/avocado/job-results/job-2019-05-06T13.46-b9bcf5f/job.log
 (1/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb30_controller.nec_xhci_model_hub.pci-root: PASS (70.02 s)
 (2/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb30_controller.qemu_xhci_model.pci-root: PASS (67.10 s)
 (3/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.nec_xhci_model_hub.pci-root: PASS (83.83 s)
 (4/4) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb30_controller.qemu_xhci_model.pci-root: PASS (81.94 s)
```
Are these cases designed for host usb disk? If so, it is better to note that in the code comments in case of confusion.

**Comment by brucepatricklee** [comment]
@qiankehan yes, the comment "please insert a usb disk into the host machine before test" has been added after your first comment

**Comment by qiankehan** [comment]
@brucepatricklee OK. No problem.

File: libvirt/tests/src/usb_device.py [hunk 4]
```diff
@@ -216,18 +228,39 @@ def usb_device_check(session, src_host=None):
...
+            if coldunplug:
+                if not vm.shutdown():
+                    test.fail("vm shutdown fail")
+
             for dev in hostdevs:
                 if dev.hostdev_type == "usb":
-                    virsh.detach_device(vm_name, dev.xml, flagstr="--live", debug=True, ignore_status=False)
+                    logging.debug("===== detach usb device {}".format(dev.source))
+                    if coldunplug:
+                        vmxml.del_device(dev)
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: For coldunplug, also need to use detach_device without flagstr
how about setting ( flagstr = "" if coldunplug else "--live" )

File: libvirt/tests/src/usb_device.py [hunk 5]
```diff
@@ -216,18 +228,39 @@ def usb_device_check(session, src_host=None):
...
+                    else:
+                        virsh.detach_device(vm_name, dev.xml, flagstr="--live", debug=True, ignore_status=False)
 
             # check the hostdev element in xml after detaching
+            if coldunplug:
+                vmxml.sync()
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: no need to sync if no vmxml operation before

File: libvirt/tests/src/usb_device.py (hunk 4 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: as discussed, remains unchanged

File: libvirt/tests/src/usb_device.py (hunk 5 above)
Comment (brucepatricklee) [reply libvirt/tests/src/usb_device.py]: as discussed, remains unchanged

**Pull request merged**